### ISSUE TITLE
Bug: Payload file does not exist - WinUI 2.5 packaging issue

### ISSUE THREAD

**Issue by dpaulino** [microsoft/microsoft-ui-xaml#2859]
**Describe the bug**

Unable to package my UWP app when using WinUI 2.5 prerelease. These errors show up:
![image](https://user-images.githubusercontent.com/7649589/86870098-9f5f5180-c08c-11ea-8509-4d2dedb7d43c.png)

**Steps to reproduce the bug**

1. Download this solution: https://1drv.ms/u/s!Ar4RhL72OhUDzrwOz9woYwhKfeDxjQ?e=m6A2qP 
2. Launch the `.sln` file in VS2019
3. Rebuild solution
3. Right click on `WapProjTemplate1` > Publish > Create app package
4. Select Microsoft Store as the distribution method
5. Ensure the configuration looks like this:
    ![image](https://user-images.githubusercontent.com/7649589/86875374-33ceb180-c097-11ea-86f1-b4e1c2391a1f.png)
6. Click create
7. Packaging error occurs in a few seconds.

**Expected behavior**

No packaging error occurs.

**Version Info**

NuGet package version: 

WinUI 2.5.0-prerelease.200609001

| Windows 10 version                  | Saw the problem? |
| :--------------------------------- | :-------------------- |
| Insider Build (xxxxx)              |    |
| May 2020 Update (19041)            |  Yes   |
| November 2019 Update (18363)       |    |
| May 2019 Update (18362)            |    |
| October 2018 Update (17763)        |    |
| April 2018 Update (17134)          |    |
| Fall Creators Update (16299)       |    |
| Creators Update (15063)            |    |

| Device form factor | Saw the problem? |
| :----------------- | :--------------- |
| Desktop            | Yes |
| Xbox               |  |
| Surface Hub        |  |
| IoT                |  |

**Additional context**

I have tried WinUI 2.4.2 using this very same `.sln` and there is no packaging error.

**Comment by ranjeshj** [comment]
@dpaulino It looks like you are using a separate wap project instead of packaging the UWP app project by itself. Is that intentional ? 
The difference between 2.4.2 and pre-release is that 2.4.2 is a framework package release, so the app includes only a link to the framework package as opposed to pre-release package which needs all the bits packaged into the app.

**Comment by dpaulino** [comment]
>Is that intentional?

Yes. Nightingale a separate wap project, so that's how I discovered this packaging error.

>The difference between 2.4.2 and pre-release is that 2.4.2 is a framework package release, so the app includes only a link to the framework package as opposed to pre-release package which needs all the bits packaged into the app.

I don't quite follow. Are you saying that the issue only occurs because 2.5 is a pre-release? And therefore, once 2.5 graduates from prerelease, this issue will no longer occur?

Or are you saying that starting with 2.5, this issue will always occur?

**Comment by chingucoding** [comment]
> I don't quite follow. Are you saying that the issue only occurs because 2.5 is a pre-release? And therefore, once 2.5 graduates from prerelease, this issue will no longer occur?
> Or are you saying that starting with 2.5, this issue will always occur?

This happens because 2.5 is a prerelase. Once 2.5 has "normal" releases, those release builds would work like 2.4 does now.

**Comment by dpaulino** [comment]
@chingucoding Is this a recent change? I have used 2.3-prerelease in the past with my app, and I never encountered this packaging error 🤔

**Comment by chingucoding** [comment]
I think that that packaging behavior @ranjeshj described was already there before 2.5, the terminal project can only use prerelease packages as they can't rely on the "normal" release packaging mechanism.

If 2.3 prereleases work for you but the 2.5 prerelease don't it sounds more like a regression. Does using a 2.3 prerelease now also work? I've had issues with VS and building projects recently, so maybe it is something in that area.

**Comment by ranjeshj** [comment]
It is a difference between prerelease and stable releases in how WinUI gets packaged. For stable releases the WinUI bits are packaged into a framework package and submitted to the store - So your app just has to say in the manifest that it depends on the WinUI framework package and not package the bits in the app. 

I would like to understand if this worked before and is not now in the wap project scenario. It is possible that something regressed - we are not specifically testing this scenario.

**Comment by dpaulino** [comment]
According to my commits, I used a pre-release package starting from november 2019 until april 2020:

![image](https://user-images.githubusercontent.com/7649589/86977467-85c41580-c131-11ea-8590-d2e7b45a0e52.png)


![image](https://user-images.githubusercontent.com/7649589/86977561-b73ce100-c131-11ea-8e12-f48488fd47ce.png)

@ranjeshj you mentioned that this may be an issue with how the bits are packaged. Is there a way for me to get a pre-release package that is packaged like a stable release? Maybe I can grab the nupkg file directly from your team's Azure Pipelines and manually add it into my project and my CI/CD pipeline? Would that be a possible workaround?

**Comment by dpaulino** [comment]
And by the way, since this issue seems related to the pre-release nature of the package, I would not want the winui team to make significant changes just to make things work for my situation. Especially since it seems this issue would eventually be resolved once WinUI 2.5 is stable. 

I would just like a workaround that lets me package successfully for now.

**Comment by stevenbrix** [comment]
@dpaulino can you share binlogs (run `msbuild /bl` to build the .wapproj) and share those? If you have one from when it works and when it doesn't that would be very helpful

**Comment by dpaulino** [comment]
@stevenbrix are these what you're looking for? 
broken: https://1drv.ms/t/s!Ar4RhL72OhUDzrwQ9q6KU09-ZbTayQ?e=HbI7Fm
successful: https://1drv.ms/t/s!Ar4RhL72OhUDzrwRNbI1ENbQ5U-mWw?e=8VGfeJ 

Also, I have working repro steps in the original issue post if you need to perform any further analysis with the msbuild process:
![image](https://user-images.githubusercontent.com/7649589/87183457-08b4af80-c29b-11ea-9484-e6ffebe108e8.png)

**Comment by stevenbrix** [comment]
@dpaulino almost, those are just plain text logs, which aren't as easy to understand. I have no problem trying out your repro next week (thanks for that), but for future reference:

In order to capture a binlog in VS:
1. Download the **[VS Project System Tools extension](https://marketplace.visualstudio.com/items?itemName=VisualStudioProductTeam.ProjectSystemTools)**
2. Set the **Build Log File** verbosity to Diagnostics: `Tools->Options->Projects and Solutions->MSBuild project build log file verbosity`
![image](https://user-images.githubusercontent.com/9649518/87231826-3bc87300-c36f-11ea-815e-d5b2b0eb8b50.png)
3. Go to View->Other Windows->Build Logging:
![image](https://user-images.githubusercontent.com/9649518/87231853-61557c80-c36f-11ea-8de4-2c974965c1ed.png)
4. Then press the little Play button to start taking logs:
![image](https://user-images.githubusercontent.com/9649518/87231869-792d0080-c36f-11ea-828b-6e13821a2c8d.png)
5. Run through the repro
      - You should see one of the builds with the status of "Failure". The file has the extension `.binlog`.

**Comment by dpaulino** [comment]
@stevenbrix thanks for the instructions. Here is the failed binlog file: https://1drv.ms/u/s!Ar4RhL72OhUDzrwViE0iqaUQU5fTuA?e=cHLuRC

**Comment by stevenbrix** [comment]
@dpaulino I'm looking into this now. Out of curiosity, why are you using the .wapproj to package your UWP app?

**Comment by dpaulino** [comment]
@stevenbrix My app package contains a .NET Framework console project that runs a console command (checknetisolation) in certain scenarios. From what I understand, UWP cannot call this console command on its own, so that's why I added a .NET framework console project, and thus I needed to use .wapproj

**Comment by dpaulino** [comment]
For more context, here's a screenshot of how I use the .wapproj project 
![image](https://user-images.githubusercontent.com/7649589/87590230-4d758780-c69b-11ea-86f6-866fc74515ef.png)

**Comment by stevenbrix** [comment]
Gotcha, that makes sense.

Found a cool little trick to build this from the command line, which really helps gather binlogs (sadly the ones produced by building from VS aren't quite as good). The key is to set the `BuildAppxUploadPackageForUap` property to `true`. So you can use a command like this to repro:

`msbuild /p:AppxBundlePlatforms=x86 /p:Platform=x86 /p:Configuration=Release /p:BuildAppxUploadPackageForUap=true /bl`

Can you try adding this workaround to the bottom of your .wapproj and see if that fixes it:


  <Target Name="_ExpandPriUploadFiles"
          Condition="'$(BuildAppxUploadPackageForUap)' == 'true'">

    <ItemGroup>
      <_PriUploadFilesToExpand Include="@(_PriFilesFromPayload)"
                               Condition="'%(OutputGroup)' != 'ProjectPriUploadFile'
                                      and '%(OutputGroup)' != 'ProjectPriFile'
                                      and '%(OutputGroup)' != 'SDKRedistOutputGroup'"
                               />
    </ItemGroup>

    <ExpandPriContent Inputs="@(_PriUploadFilesToExpand)"
                      MakePriExeFullPath="$(MakePriExeFullPath)"
                      MakePriExtensionPath="$(OutOfProcessMakePriExtensionPath)"
                      IntermediateDirectory="$(IntermediateOutputPath)"
                      AdditionalMakepriExeParameters="$(AppxExpandPriContentAdditionalMakepriExeParameters)"
                      ExcludeXamlFromLibraryLayoutsWhenXbfIsPresent="$(AppxExcludeXamlFromLibraryLayoutsWhenXbfIsPresent)"
                      VsTelemetrySession="$(VsTelemetrySession)"
                          >
      <Output TaskParameter="Expanded" ItemName="_ExpandedPriUploadPayload" />
      <Output TaskParameter="IntermediateFileWrites" ItemName="FileWrites" />
    </ExpandPriContent>
  </Target>

**Comment by dpaulino** [comment]
@stevenbrix The workaround worked for the test project! I'll try this out on my real app now. 

Do you think this is an issue with the .wapproj template? If so, I can try to talk to someone on the packaging team

**Comment by stevenbrix** [comment]
It's an issue somewhere in there for sure. That target I copy/pasted from Microsoft.AppxPackaging.targets and added the following line:

`ExcludeXamlFromLibraryLayoutsWhenXbfIsPresent="$(AppxExcludeXamlFromLibraryLayoutsWhenXbfIsPresent)"`

This tells the `ExpandPriContent` task to ignore the .xaml files when the .xbf files are present.

It's probably worth filing a feedback issue through VS with the logs for the repro (and this information as well).

/cc @svetbonev

**Comment by dpaulino** [comment]
@stevenbrix The workaround worked for my real app! Can't thank you enough!! 🙌🏽🎉🥳

**Comment by stevenbrix** [comment]
@dpaulino that's awesome, glad you're unblocked 😄 

I've filed this internal bug on VS for this issue: https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1157568
